File: android/src/main/kotlin/dev/takzobye/flutter_social_share_plus/InstagramShareHandler.kt
package dev.takzobye.flutter_social_share_plus

import android.app.Activity
import android.content.ClipData
import android.content.Intent
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodChannel.Result

class InstagramShareHandler internal constructor(
    private val mediaPreparer: MediaPreparer = ShareMediaPreparer,
    private val activityRequests: ActivityRequestScope<Activity> = ActivityRequestScope(),
    private val mediaAvailability: MediaAvailability = ShareMediaAvailability,
) {
    fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityRequests.attach(binding.activity)
    }

    fun onDetachedFromActivityForConfigChanges() {
        activityRequests.detach(ShareResponse.activityRecreated())
    }

    fun onDetachedFromActivity() {
        activityRequests.detach(ShareResponse.activityDetached())
    }

    fun onDetachedFromEngine() {
        onDetachedFromActivity()
    }

    fun shareFeed(activity: Activity, filePath: String?, result: Result) {
        val path = filePath?.takeIf { it.isNotBlank() }
        if (path == null) {
            result.success(ShareResponse.failed("invalid_input", "A media path is required"))
            return
        }

        val intent = Intent(Intent.ACTION_SEND).apply { setPackage(INSTAGRAM_PACKAGE) }
        if (!listOf("image/*", "video/*").any { type ->
                mediaAvailability.canHandle(activity, Intent(intent).apply { setType(type) })
            }) {
            result.success(ShareResponse.unavailable())
            return
        }

        val request = activityRequests.begin { response -> result.success(response) }
        if (request == null) {
            result.success(ShareResponse.failed("platform_error", "Activity not available"))
            return
        }

        try {
            mediaPreparer.prepareAsync(
                context = activity.applicationContext,
                paths = listOf(path),
                maxVideoBytes = null,
            ) { media, error ->
                val currentActivity = activityRequests.activityFor(request)
                    ?: return@prepareAsync
                if (error != null) {
                    activityRequests.complete(request, error)
                    return@prepareAsync
                }
                val prepared = media?.singleOrNull()
                if (prepared == null) {
                    activityRequests.complete(
                        request,
                        ShareResponse.failed("platform_error", "Unable to prepare media"),
                    )
                    return@prepareAsync
                }
                if (!prepared.mimeType.startsWith("image/") &&
                    !prepared.mimeType.startsWith("video/")
                ) {
                    activityRequests.complete(
                        request,
                        ShareResponse.failed(
                            "unsupported_media",
                            "Instagram requires an image or video",
                        ),
                    )
                    return@prepareAsync
                }

                val shareIntent = Intent(intent).apply {
                    type = prepared.mimeType
                    putExtra(Intent.EXTRA_STREAM, prepared.uri)
                    clipData = ClipData.newRawUri("shared media", prepared.uri)
                    addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                }
                try {
                    ShareMedia.grant(currentActivity, prepared, INSTAGRAM_PACKAGE)
                    currentActivity.startActivity(shareIntent)
                    activityRequests.complete(request, ShareResponse.launched())
                } catch (error: Exception) {
                    activityRequests.complete(
                        request,
                        ShareResponse.failed(
                            "platform_error",
                            error.message ?: "Unable to open Instagram",
                        ),
                    )
                }
            }
        } catch (error: Exception) {
            activityRequests.complete(
                request,
                ShareResponse.failed(
                    "platform_error",
                    error.message ?: "Unable to prepare media",
                ),
            )
        }
    }

    fun shareStory(
        activity: Activity,
        appId: String?,
        stickerPath: String?,
        backgroundImagePath: String?,
        backgroundVideoPath: String?,
        backgroundTopColor: String?,
        backgroundBottomColor: String?,
        attributionUrl: String?,
        result: Result,
    ) {
        if (appId.isNullOrBlank()) {
            result.success(ShareResponse.failed("invalid_input", "A Facebook App ID is required"))
            return
        }
        if (backgroundImagePath != null && backgroundVideoPath != null) {
            result.success(
                ShareResponse.failed(
                    "invalid_input",
                    "A Story cannot contain both an image and a video background",
                )
            )
            return
        }
        val paths = listOfNotNull(stickerPath, backgroundImagePath, backgroundVideoPath)
        if (paths.isEmpty()) {
            result.success(ShareResponse.failed("invalid_input", "Story content is required"))
            return
        }

        val intent = Intent(INSTAGRAM_STORY_ACTION).apply {
            setPackage(INSTAGRAM_PACKAGE)
        }
        if (!mediaAvailability.canHandle(activity, intent)) {
            result.success(ShareResponse.unavailable())
            return
        }

        val request = activityRequests.begin { response -> result.success(response) }
        if (request == null) {
            result.success(ShareResponse.failed("platform_error", "Activity not available"))
            return
        }

        try {
            mediaPreparer.prepareAsync(
                context = activity.applicationContext,
                paths = paths,
                maxVideoBytes = MAX_STORY_VIDEO_BYTES,
            ) { media, error ->
                val currentActivity = activityRequests.activityFor(request)
                    ?: return@prepareAsync
                if (error != null) {
                    activityRequests.complete(request, error)
                    return@prepareAsync
                }
                val prepared = media ?: run {
                    activityRequests.complete(
                        request,
                        ShareResponse.failed("platform_error", "Unable to prepare Story"),
                    )
                    return@prepareAsync
                }
                val sticker = stickerPath?.let { prepared[paths.indexOf(it)] }
                val backgroundImage = backgroundImagePath?.let { prepared[paths.indexOf(it)] }
                val backgroundVideo = backgroundVideoPath?.let { prepared[paths.indexOf(it)] }
                if (sticker != null && !sticker.mimeType.startsWith("image/")) {
                    activityRequests.complete(
                        request,
                        ShareResponse.failed("unsupported_media", "Sticker must be an image"),
                    )
                    return@prepareAsync
                }
                if (backgroundImage != null && !backgroundImage.mimeType.startsWith("image/")) {
                    activityRequests.complete(
                        request,
                        ShareResponse.failed("unsupported_media", "Background must be an image"),
                    )
                    return@prepareAsync
                }
                if (backgroundVideo != null && !backgroundVideo.mimeType.startsWith("video/")) {
                    activityRequests.complete(
                        request,
                        ShareResponse.failed("unsupported_media", "Background must be a video"),
                    )
                    return@prepareAsync
                }
                val shareIntent = Intent(intent).apply {
                    putExtra("source_application", appId)
                    backgroundImage?.let { mediaFile ->
                        data = mediaFile.uri
                        type = mediaFile.mimeType
                        clipData = ClipData.newRawUri("background", mediaFile.uri)
                    }
                    backgroundVideo?.let { mediaFile ->
                        data = mediaFile.uri
                        type = mediaFile.mimeType
                        clipData = ClipData.newRawUri("background", mediaFile.uri)
                    }
                    sticker?.let { mediaFile -> putExtra("interactive_asset_uri", mediaFile.uri) }
                    backgroundTopColor?.let { putExtra("top_background_color", it) }
                    backgroundBottomColor?.let { putExtra("bottom_background_color", it) }
                    attributionUrl?.let { putExtra("content_url", it) }
                    addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                }
                try {
                    backgroundImage?.let { ShareMedia.grant(currentActivity, it, INSTAGRAM_PACKAGE) }
                    backgroundVideo?.let { ShareMedia.grant(currentActivity, it, INSTAGRAM_PACKAGE) }
                    sticker?.let { ShareMedia.grant(currentActivity, it, INSTAGRAM_PACKAGE) }
                    currentActivity.startActivity(shareIntent)
                    activityRequests.complete(request, ShareResponse.launched())
                } catch (error: Exception) {
                    activityRequests.complete(
                        request,
                        ShareResponse.failed(
                            "platform_error",
                            error.message ?: "Unable to open Instagram Stories",
                        ),
                    )
                }
            }
        } catch (error: Exception) {
            activityRequests.complete(
                request,
                ShareResponse.failed(
                    "platform_error",
                    error.message ?: "Unable to prepare Instagram Stories",
                ),
            )
        }
    }

    private companion object {
        const val INSTAGRAM_PACKAGE = "com.instagram.android"
        const val INSTAGRAM_STORY_ACTION = "com.instagram.share.ADD_TO_STORY"
        const val MAX_STORY_VIDEO_BYTES = 50L * 1024 * 1024
    }
}

File: ios/flutter_social_share_plus/Sources/flutter_social_share_plus/InstagramShareHandler.swift
import Flutter
import Photos
import UIKit

final class InstagramShareHandler {
    func isFeedAvailable() -> Bool {
        guard let url = URL(string: "instagram://") else { return false }
        return UIApplication.shared.canOpenURL(url)
    }

    func isStoryAvailable() -> Bool {
        guard let url = URL(string: "instagram-stories://share") else { return false }
        return UIApplication.shared.canOpenURL(url)
    }

    func shareFeed(filePath: String?, result: @escaping FlutterResult) {
        guard isFeedAvailable() else {
            result(ShareResponse.unavailable())
            return
        }
        guard let path = filePath else {
            result(ShareResponse.failed("invalid_input", "A media path is required"))
            return
        }
        let validation = ShareMedia.validatePath(
            path,
            allowed: ["image/", "video/"]
        )
        guard case let .success(fileURL) = validation else {
            if case let .failure(error) = validation {
                result(error.response)
            }
            return
        }

        requestPhotoLibraryAccess { granted in
            guard granted else {
                result(ShareResponse.failed("permission_denied", "Photo library access was denied"))
                return
            }
            self.saveAndOpen(fileURL: fileURL, result: result)
        }
    }

    func shareStory(
        appId: String?,
        stickerPath: String?,
        backgroundImagePath: String?,
        backgroundVideoPath: String?,
        backgroundTopColor: String?,
        backgroundBottomColor: String?,
        attributionUrl: String?,
        result: @escaping FlutterResult
    ) {
        guard let appId, !appId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            result(ShareResponse.failed("invalid_input", "A Facebook App ID is required"))
            return
        }
        guard backgroundImagePath == nil || backgroundVideoPath == nil else {
            result(ShareResponse.failed("invalid_input", "A Story cannot contain both backgrounds"))
            return
        }
        let paths = [stickerPath, backgroundImagePath, backgroundVideoPath].compactMap { $0 }
        guard !paths.isEmpty else {
            result(ShareResponse.failed("invalid_input", "Story content is required"))
            return
        }
        guard let storyURL = URLComponents(
            string: "instagram-stories://share"
        ).flatMap({ components in
            var components = components
            components.queryItems = [URLQueryItem(name: "source_application", value: appId)]
            return components.url
        }), UIApplication.shared.canOpenURL(storyURL) else {
            result(ShareResponse.unavailable())
            return
        }

        for path in paths {
            let allowed: Set<String> = path == backgroundVideoPath ? ["video/"] : ["image/"]
            let validation = ShareMedia.validatePath(path, allowed: allowed)
            guard case .success = validation else {
                if case let .failure(error) = validation {
                    result(error.response)
                }
                return
            }
        }

        ShareMedia.load(paths: paths, maxVideoBytes: 50 * 1024 * 1024) { loaded in
            guard case let .success(data) = loaded else {
                if case let .failure(error) = loaded { result(error.response) }
                return
            }
            var item = [String: Any]()
            if let attributionUrl {
                // contentURL is the Story pasteboard key used by Meta's historical examples.
                item["com.instagram.sharedSticker.contentURL"] = attributionUrl
            }
            if let backgroundImagePath {
                item["com.instagram.sharedSticker.backgroundImage"] = data[URL(fileURLWithPath: backgroundImagePath)]
            }
            if let backgroundVideoPath {
                item["com.instagram.sharedSticker.backgroundVideo"] = data[URL(fileURLWithPath: backgroundVideoPath)]
            }
            if let stickerPath {
                item["com.instagram.sharedSticker.stickerImage"] = data[URL(fileURLWithPath: stickerPath)]
            }
            if let backgroundTopColor {
                item["com.instagram.sharedSticker.backgroundTopColor"] = backgroundTopColor
            }
            if let backgroundBottomColor {
                item["com.instagram.sharedSticker.backgroundBottomColor"] = backgroundBottomColor
            }

            UIPasteboard.general.setItems(
                [item],
                options: [
                    .expirationDate: Date().addingTimeInterval(300),
                    .localOnly: true,
                ]
            )
            UIApplication.shared.open(storyURL) { opened in
                result(opened ? ShareResponse.launched() : ShareResponse.failed(
                    "platform_error",
                    "Unable to open Instagram Stories"
                ))
            }
        }
    }

    private func saveAndOpen(fileURL: URL, result: @escaping FlutterResult) {
        var localIdentifier: String?
        let isVideo = ShareMedia.mimeType(for: fileURL)?.hasPrefix("video/") == true

        PHPhotoLibrary.shared().performChanges({
            let request = isVideo
                ? PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: fileURL)
                : PHAssetChangeRequest.creationRequestForAssetFromImage(atFileURL: fileURL)
            localIdentifier = request?.placeholderForCreatedAsset?.localIdentifier
        }) { success, error in
            DispatchQueue.main.async {
                guard success, let localIdentifier else {
                    result(ShareResponse.failed(
                        "platform_error",
                        error?.localizedDescription ?? "Unable to save media"
                    ))
                    return
                }
                var components = URLComponents(string: "instagram://library")
                components?.queryItems = [URLQueryItem(name: "LocalIdentifier", value: localIdentifier)]
                guard let instagramURL = components?.url else {
                    result(ShareResponse.failed("platform_error", "Unable to build Instagram URL"))
                    return
                }
                UIApplication.shared.open(instagramURL) { opened in
                    result(opened ? ShareResponse.launched() : ShareResponse.failed(
                        "platform_error",
                        "Unable to open Instagram"
                    ))
                }
            }
        }
    }

    private func requestPhotoLibraryAccess(completion: @escaping (Bool) -> Void) {
        switch PHPhotoLibrary.authorizationStatus(for: .addOnly) {
        case .authorized, .limited:
            completion(true)
        case .notDetermined:
            PHPhotoLibrary.requestAuthorization(for: .addOnly) { status in
                DispatchQueue.main.async {
                    completion(status == .authorized || status == .limited)
                }
            }
        default:
            completion(false)
        }
    }
}
